toolchain/check/handle_binding_pattern.cpp

```diff
@@ -54,13 +46,13 @@ static auto HandleAnyBindingPattern(Context& context, Parse::NodeId node_id,
     auto entity_name_id = context.entity_names().Add(
         {.name_id = name_id,
          .parent_scope_id = context.scope_stack().PeekNameScopeId(),
-         // TODO: Don't allocate a compile-time binding index for an associated
-         // constant declaration.
-         .bind_index = needs_compile_time_binding
-                           ? context.scope_stack().AddCompileTimeBinding()
-                           : SemIR::CompileTimeBindIndex::None});
+         .bind_index_value =
+             (is_generic ? context.scope_stack().AddCompileTimeBinding()
+                         : SemIR::CompileTimeBindIndex::None)
+                 .index,
```

Comment:
Would it make sense to switch to a constructor, to encapsulate away the `.index`?

Reply:
Encapsulated into a distinct `Add` function on the value store. (This gets us slightly closer to using distinct storage for this case, and also keeps the field names for the remaining cases.)

Reply:
Sure, SG, particularly since we're talking about refactoring anyways.